contribs/gnodev/cmd/gnodev/main.go

```diff
@@ -0,0 +1,356 @@
+	tmlog "github.com/gnolang/gno/tm2/pkg/log"
+	osm "github.com/gnolang/gno/tm2/pkg/os"
+)
+
+type devCfg struct {
+	bindAddr string
+	minimal  bool
+	verbose  bool
+	noWatch  bool
+}
+
+var defaultDevOptions = &devCfg{
+	bindAddr: "127.0.0.1:8888",
+}
+
+func main() {
+	cfg := &devCfg{}
+
+	stdio := commands.NewDefaultIO()
+	cmd := commands.NewCommand(
+		commands.Metadata{
+			Name:       "gnodev",
+			ShortUsage: "gnodev [flags] <path>",
+			ShortHelp:  "GnoDev run a node for dev purpose, it will load the given package path",
+		},
+		cfg,
+		func(_ context.Context, args []string) error {
+			return execDev(cfg, args, stdio)
+		})
+
+	if err := cmd.ParseAndRun(context.Background(), os.Args[1:]); err != nil {
+		fmt.Fprintf(os.Stderr, "%+v\n", err)
+		os.Exit(1)
+	}
+}
+
+func (c *devCfg) RegisterFlags(fs *flag.FlagSet) {
+	fs.StringVar(
+		&c.bindAddr,
+		"web-bind",
+		defaultDevOptions.bindAddr,
+		"verbose output when deving",
+	)
+
+	fs.BoolVar(
+		&c.minimal,
+		"minimal",
+		defaultDevOptions.verbose,
+		"don't load example folder packages along default transaction",
+	)
+
+	fs.BoolVar(
+		&c.verbose,
+		"verbose",
+		defaultDevOptions.verbose,
+		"verbose output when deving",
+	)
+
+	fs.BoolVar(
+		&c.noWatch,
+		"no-watch",
+		defaultDevOptions.noWatch,
+		"watch for files change",
+	)
+
+}
+
+func execDev(cfg *devCfg, args []string, io commands.IO) error {
+	ctx, cancel := context.WithCancelCause(context.Background())
+	defer cancel(nil)
+
+	// guess root dir
+	gnoroot := gnoenv.RootDir()
+
+	pkgpaths, err := parseArgsPackages(io, args)
+	if err != nil {
+		return fmt.Errorf("unable to parse package paths: %w", err)
+	}
+
+	// noopLogger := log.NewTMLogger(log.NewSyncWriter(io.Out))
+	noopLogger := tmlog.NewNopLogger()
+
+	// RAWTerm setup
+	rt := gnodev.NewRawTerm()
+	{
+		restore, err := rt.Init()
+		if err != nil {
+			return fmt.Errorf("unable to init raw term for dev: %s", err)
+		}
+		defer restore()
+
+		// correctly format output for terminal
+		io.SetOut(commands.WriteNopCloser(rt))
+
+		// Setup trap signal
+		osm.TrapSignal(func() {
+			restore()
+			cancel(nil)
+		})
+	}
+
+	nodeOut := rt.NamespacedWriter("Node")
+	webOut := rt.NamespacedWriter("GnoWeb")
+	keyOut := rt.NamespacedWriter("KeyPress")
+
+	var dnode *gnodev.Node
+	{
+		var err error
+		// XXX: redirect node to output file
+		dnode, err = setupDevNode(ctx, noopLogger, cfg, pkgpaths, gnoroot)
+		if err != nil {
+			return err // already formated in setupDevNode
+		}
+		defer dnode.Close()
+	}
+
+	fmt.Fprintf(nodeOut, "Listener: %s\n", dnode.GetRemoteAddress())
+
+	// setup files watcher
+	w, err := setupPkgsWatcher(cfg, dnode.ListPkgs())
+	if err != nil {
+		return fmt.Errorf("unable to watch for files change: %w", err)
+	}
+
+	ccpath := make(chan []string, 1)
+	go func() {
+		defer close(ccpath)
+
+		const debounceTimeout = time.Millisecond * 500
+
+		if err := handleDebounce(ctx, w, ccpath, debounceTimeout); err != nil {
```

Comment:
What do you think about moving this go routine logic inside `handleDebounce`, and have it return a channel (ccpath), instead of an error? The only error you return from this method is for the context, which doesn't make sense (you recancel the same context with the ctx.Error()) here. I would drop error handling completely.

This way, you can lower the amount of args (drop the channel, and timeout, which is a constant and can live inside `handleDebounce`)

Reply:
You probably missed that the method can also return errors from `<-watcher.Errors`. That's why I handle this here instead of wrapping everything in `handleDebounce`. I can still move the constant inside it.